### PR TITLE
HW3: Ali Ammar

### PR DESCRIPTION
**Pull request by ammarali32**

### CODE REVIEW

File: src/main/java/api/weather/YearTemperatureStats.java [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package api.weather;
 
+
```
Comment (Walingar) [inline src/main/java/api/weather/YearTemperatureStats.java]: Why did you add this line?

File: src/main/java/impl/weather/DayTemperatureInfoImpl.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import api.weather.DayTemperatureInfo;
+import kotlin.Pair;
+import org.jetbrains.annotations.NotNull;
+
+import java.time.Month;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+
+public class DayTemperatureInfoImpl implements DayTemperatureInfo {
+    private int day;
```
Comment (Walingar) [inline src/main/java/impl/weather/DayTemperatureInfoImpl.java]: final

File: src/main/java/impl/weather/tempInfoComp.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package impl.weather;
+
+import api.weather.DayTemperatureInfo;
+
+import java.util.Comparator;
+
+public class tempInfoComp implements Comparator<DayTemperatureInfo> {
```
Comment (Walingar) [inline src/main/java/impl/weather/tempInfoComp.java]: codestyle

File: src/main/java/impl/weather/tempInfoComp.java [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    @Override
+    public int compare(DayTemperatureInfo o1, DayTemperatureInfo o2) {
+        return o1.getTemperature() - o2.getTemperature() - 1;
```
Comment (Walingar) [inline src/main/java/impl/weather/tempInfoComp.java]: Why do you subtract -1 here?

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+import api.weather.DayTemperatureInfo;
+import api.weather.YearTemperatureStats;
+import kotlin.Pair;
+
+import java.awt.*;
+import java.time.Month;
+import java.util.*;
+import java.util.List;
+
+public class YearTemperatureStatsimpl implements YearTemperatureStats {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: codestyle

File: src/main/java/impl/weather/YearTemperatureStatsParserimpl.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import api.weather.DayTemperatureInfo;
+import api.weather.YearTemperatureStats;
+import api.weather.YearTemperatureStatsParser;
+
+import java.sql.SQLSyntaxErrorException;
+import java.time.Month;
+import java.util.Collection;
+
+public class YearTemperatureStatsParserimpl implements YearTemperatureStatsParser {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsParserimpl.java]: codestyle

File: src/main/java/impl/weather/YearTemperatureStatsParserimpl.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+    @Override
+    public YearTemperatureStats parse(Collection<String> rawData) {
+        YearTemperatureStats status = new YearTemperatureStatsimpl();
+        for (String input : rawData) {
+            var split = input.split("[. ]");
+            var day = Integer.parseInt(split[0]);
+            var month = Month.of(Integer.parseInt(split[1]));
+            var temp = Integer.parseInt(split[2]);
+            DayTemperatureInfo newday = new DayTemperatureInfoImpl(day, month, temp);
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsParserimpl.java]: codestyle

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    private HashMap<Month, HashMap<Integer, DayTemperatureInfo>> status = new HashMap<>();
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: final

Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: Use interface here. We don't use any specific for HashMap methods here

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    private HashMap<Month, Pair<Double, Integer>> averageTemperature = new HashMap<>();
+    private HashMap<Month, Integer> maximumTemperature = new HashMap<>();
+
+    @Override
+    public void updateStats(DayTemperatureInfo info) {
+        int day = info.getDay();
+        Month month = info.getMonth();
+        int temperature = info.getTemperature();
+        if (status.containsKey(month) == false) {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: `!status.containsKey(month)`

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+import kotlin.Pair;
+
+import java.awt.*;
+import java.time.Month;
+import java.util.*;
+import java.util.List;
+
+public class YearTemperatureStatsimpl implements YearTemperatureStats {
+    private HashMap<Month, HashMap<Integer, DayTemperatureInfo>> status = new HashMap<>();
+    private HashMap<Month, Pair<Double, Integer>> averageTemperature = new HashMap<>();
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: Please, try not to use Pair. It is better to create separate class which present such data

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    private HashMap<Month, Integer> maximumTemperature = new HashMap<>();
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: It is better to create separate class which hold and update avg, max etc.

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+            status.put(month, new HashMap<>());
+        }
+        status.get(month).put(day, info);
+        var average = averageTemperature.getOrDefault(month, null);
+        if (average != null) {
+            double numOfDays = (average.component2());
+            var oldaverage = average.component1();
+            double newAverage = ((oldaverage * numOfDays) + temperature) / (numOfDays + 1);
+            int counter = (int) numOfDays + 1;
+            averageTemperature.replace(month, average, new Pair(newAverage, counter));
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: Pair<>

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        if (max != null) {
+            if (max < temperature) {
+                maximumTemperature.replace(month, max, temperature);
+            }
+        } else {
+            maximumTemperature.put(month, temperature);
+        }
+    }
+
+    public boolean cmp(int a, int b) {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: Unused method

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    public Map<Month, Integer> getMaxTemperature() {
+        return maximumTemperature;
+    }
+
+    @Override
+    public List<DayTemperatureInfo> getSortedTemperature(Month month) {
+        if (status.getOrDefault(month, null) == null) {
+            return new ArrayList<>();
+        }
+        HashMap<Integer, DayTemperatureInfo> temporary = status.getOrDefault(month, null);
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: interface

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+        ArrayList<DayTemperatureInfo> List = new ArrayList<DayTemperatureInfo>(temporary.values());
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: ArrayList<>

Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: codestyle

File: src/main/java/impl/weather/YearTemperatureStatsimpl.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+        Collections.sort(List, new tempInfoComp());
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsimpl.java]: List.sort

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/weather/DayTemperatureInfoImpl.java (hunk 2 above)
Comment (ammarali32) [reply src/main/java/impl/weather/DayTemperatureInfoImpl.java]: What do you mean by final ??

File: src/main/java/impl/weather/tempInfoComp.java (hunk 3 above)
Comment (ammarali32) [reply src/main/java/impl/weather/tempInfoComp.java]: What do you mean ?? :(

File: src/main/java/impl/weather/DayTemperatureInfoImpl.java (hunk 2 above)
Comment (Walingar) [reply src/main/java/impl/weather/DayTemperatureInfoImpl.java]: you can mark a field or variable as final, then link to this field won't be modified. It is good practice to mark such fields as final

File: src/main/java/impl/weather/tempInfoComp.java (hunk 4 above)
Comment (ammarali32) [reply src/main/java/impl/weather/tempInfoComp.java]: Because of the way the comp works it will swap the places if they are equal and as long as we don't want that to happen we should subtract a small value eps to keep it working fine but since we are working with integers we can subtract 1.

File: src/main/java/impl/weather/tempInfoComp.java (hunk 3 above)
Comment (Walingar) [reply src/main/java/impl/weather/tempInfoComp.java]: Read code conventions, please, https://www.oracle.com/technetwork/java/codeconvtoc-136057.html

File: src/main/java/impl/weather/tempInfoComp.java (hunk 4 above)
Comment (Walingar) [reply src/main/java/impl/weather/tempInfoComp.java]: Did you mean how sorting works? Comparator just returns an integer no more. The default sorting algorithm in Java is merge sort, so it is a stable sorting algorithm

Comment (Walingar) [reply src/main/java/impl/weather/tempInfoComp.java]: You saw the wrong order due to another problem in your code - using HashMap

File: src/main/java/impl/weather/tempInfoComp.java (hunk 3 above)
Comment (ammarali32) [reply src/main/java/impl/weather/tempInfoComp.java]: thanks but I still don't know what is the problem with these lines ?? and others? could you please just specify the problem

Comment (ammarali32) [reply src/main/java/impl/weather/tempInfoComp.java]: ???

Comment (Walingar) [reply src/main/java/impl/weather/tempInfoComp.java]: Oh, excuse me, I missed it.  https://www.oracle.com/technetwork/java/javase/documentation/codeconventions-135099.html#367

Comment (ammarali32) [reply src/main/java/impl/weather/tempInfoComp.java]: thanks ))

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+import api.weather.DayTemperatureInfo;
+import api.weather.YearTemperatureStats;
+
+import java.time.Month;
+import java.util.*;
+import java.util.List;
+
+public class YearTemperatureStatsImpl implements YearTemperatureStats {
+    private final Map<Month, Map<Integer, DayTemperatureInfo>> status = new HashMap<>();
+    private final Average average = new Average();
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: I meant that it would be better to create one class that contains information about the month, rather than three classes, each of which contains a map

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Map<Month, Integer> getMaxTemperature() {
+        return this.maxTemperature.getMaxTemperature();
+    }
+
+    @Override
+    public List<DayTemperatureInfo> getSortedTemperature(Month month) {
+        if (status.getOrDefault(month, null) == null) {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Doyou need getOrDefault here?

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            return new ArrayList<>();
+        }
+        Map<Integer, DayTemperatureInfo> temperatureInfo;
+        temperatureInfo = status.getOrDefault(month, null);
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: You did getOrDefault before

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        ArrayList<DayTemperatureInfo> sortedTemperature = new ArrayList<>(temperatureInfo.values());
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Interface

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/weather/DayTemperatureInfoComperator.java [hunk 21]
```diff
@@ -0,0 +1,15 @@
+package impl.weather;
+
+import api.weather.DayTemperatureInfo;
+
+import java.util.Comparator;
+
+public class DayTemperatureInfoComperator implements Comparator<DayTemperatureInfo> {
```
Comment (Walingar) [inline src/main/java/impl/weather/DayTemperatureInfoComperator.java]: Comparator :)

File: src/main/java/impl/weather/Info.java [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    public void updateAverage(int temperature) {
+        double numOfDays = (this.averageInfo.getSecond());
+        var oldAverage = this.averageInfo.getFirst();
+        double newAverage = ((oldAverage * numOfDays) + temperature) / (numOfDays + 1);
+        int counter = (int) numOfDays + 1;
+        this.averageInfo = new Pair(newAverage, counter);
+    }
+
+    public Double getAverageTemp() {
+        var information = this.averageInfo;
```
Comment (Walingar) [inline src/main/java/impl/weather/Info.java]: this is not needed

File: src/main/java/impl/weather/Info.java [hunk 23]
```diff
@@ -0,0 +1,49 @@
...
+import api.weather.DayTemperatureInfo;
+
+import java.time.Month;
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class Info {
+    private final Map<Integer, DayTemperatureInfo> info = new HashMap<>();
+    private Pair averageInfo;
```
Comment (Walingar) [inline src/main/java/impl/weather/Info.java]: Please, don't use Pair. Create a nested class which represents a data which you store here

File: src/main/java/impl/weather/Info.java [hunk 24]
```diff
@@ -0,0 +1,49 @@
+package impl.weather;
+
+import api.weather.DayTemperatureInfo;
+
+import java.time.Month;
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class Info {
```
Comment (Walingar) [inline src/main/java/impl/weather/Info.java]: Info of what? It's quite strange name

File: src/main/java/impl/weather/Pair.java [hunk 25]
```diff
@@ -0,0 +1,19 @@
+package impl.weather;
+
+public class Pair {
```
Comment (Walingar) [inline src/main/java/impl/weather/Pair.java]: It's a part of `Info` implementation, so it should be a nested class

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 26]
```diff
@@ -0,0 +1,72 @@
...
+        if (current == null) {
+            return null;
+        }
+        return this.status.get(month).getAverageTemp();
+    }
+
+    @Override
+    public Map<Month, Integer> getMaxTemperature() {
+        Map<Month, Integer> ans = new HashMap<>();
+        for (int i = 1; i <= 12; i++) {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: What if one more month will be added ?)

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 27]
```diff
@@ -0,0 +1,72 @@
...
+            Month month = Month.of(i);
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: So you can iterate over the keys of status map

**Review by Walingar — CHANGES_REQUESTED** [review]
Much better! Let's fix some mistakes

File: src/main/java/impl/weather/monthInfo.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
+package impl.weather;
+
+import api.weather.DayTemperatureInfo;
+
+import java.util.HashMap;
+import java.util.Map;
+
+
+public class monthInfo {
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: code style

File: src/main/java/impl/weather/monthInfo.java [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+    public void updateMaxValue(int temperature) {
+        this.maximumTemp = Math.max(temperature, this.maximumTemp);
+    }
+
+    public Integer getMaxTemperature() {
+        return this.maximumTemp;
+    }
+
+    public static class averageData {
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: code style

File: src/main/java/impl/weather/monthInfo.java [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+        public averageData(double average, int numOfElements) {
+            this.average = average;
+            this.numOfElements = numOfElements;
+        }
+
+        public double getAverage() {
+            return this.average;
+        }
+
+        public int getNumOfElements() {
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: Do you need it?

File: src/main/java/impl/weather/monthInfo.java [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+        private final double average;
+        private final int numOfElements;
+
+        public averageData(double average, int numOfElements) {
+            this.average = average;
+            this.numOfElements = numOfElements;
+        }
+
+        public double getAverage() {
+            return this.average;
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: this is redundant

File: src/main/java/impl/weather/monthInfo.java [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+    public Double getAverageTemp() {
+        return this.averageInfo.getAverage();
+    }
+
+    public void updateMaxValue(int temperature) {
+        this.maximumTemp = Math.max(temperature, this.maximumTemp);
+    }
+
+    public Integer getMaxTemperature() {
+        return this.maximumTemp;
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: this is redundant

File: src/main/java/impl/weather/monthInfo.java [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+        info.put(day, dayInfo);
+    }
+
+
+    public Map<Integer, DayTemperatureInfo> getInfo() {
+        return this.info;
+    }
+
+    public void updateAverage(int temperature) {
+        double numOfDays = (this.averageInfo.getNumOfElements());
```
Comment (Walingar) [inline src/main/java/impl/weather/monthInfo.java]: is it double?

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+
+public class YearTemperatureStatsImpl implements YearTemperatureStats {
+    private final Map<Month, monthInfo> status = new HashMap<>();
+
+    @Override
+    public void updateStats(DayTemperatureInfo info) {
+        int day = info.getDay();
+        Month month = info.getMonth();
+        int temperature = info.getTemperature();
+        if (!status.containsKey(month)) {
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: use putIfAbsent

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 35]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Override
+    public List<DayTemperatureInfo> getSortedTemperature(Month month) {
+        monthInfo current;
+        current = status.get(month);
+        if (current == null) {
+            return new ArrayList<>();
+        }
+        List<DayTemperatureInfo> sortedTemperature = new ArrayList<>(current.getInfo().values());
+        sortedTemperature.sort(Comparator.comparing(DayTemperatureInfo::getTemperature).thenComparing(DayTemperatureInfo::getDay, Comparator.reverseOrder()));
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: You should consider the order of days by addition. How do you do it?

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+        return sortedTemperature;
+    }
+
+    @Override
+    public DayTemperatureInfo getTemperature(int day, Month month) {
+        monthInfo current = status.get(month);
+        if (current == null) {
+            return null;
+        }
+        return current.getInfo().getOrDefault(day, null);
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Do you need this getOrDefault?

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/weather/monthInfo.java (hunk 32 above)
Comment (ammarali32) [reply src/main/java/impl/weather/monthInfo.java]: No, it is not this one find the max temperature on each month while the second one find the max temperature over the year

**Review by Walingar — CHANGES_REQUESTED** [review]
Please, remove redundant `this`

**Comment by ammarali32** [comment]
> Please, remove redundant `this`

Done ))

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 37]
```diff
@@ -0,0 +1,62 @@
...
+
+    @Override
+    public List<DayTemperatureInfo> getSortedTemperature(Month month) {
+        MonthInfo current;
+        current = status.get(month);
+        if (current == null) {
+            return new ArrayList<>();
+        }
+        return current.getInfo().values().stream()
+                .sorted(Comparator.comparingInt(DayTemperatureInfo::getTemperature).thenComparing(DayTemperatureInfo::getDay,Comparator.reverseOrder()))
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: code-style

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java [hunk 38]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public Map<Month, Integer> getMaxTemperature() {
+        Map<Month, Integer> ans = new HashMap<>();
+        status.forEach((Month, MonthInfo) -> ans.put(Month, MonthInfo.getMaxTemperatureOnMonth()));
+        return ans;
+    }
+
+    @Override
+    public List<DayTemperatureInfo> getSortedTemperature(Month month) {
+        MonthInfo current;
```
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: You can merge next line with this one

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java (hunk 37 above)
Comment (Walingar) [inline src/main/java/impl/weather/YearTemperatureStatsImpl.java]: How do you consider days order with equal temp?

**Review by Walingar — CHANGES_REQUESTED** [review]

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java (hunk 37 above)
Comment (ammarali32) [reply src/main/java/impl/weather/YearTemperatureStatsImpl.java]: As you can see the second comparator is comparing according to the day and as long as the temperature int the i + 1 day will not be entered before the temp in the i day in this case if the temp are equals in two days  then they will be sorted according to the day which mean taking the input into account

Comment (Walingar) [reply src/main/java/impl/weather/YearTemperatureStatsImpl.java]: As I see you just compare them by day number using reverse order.

Comment (ammarali32) [reply src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Yes exactly, because the day number and order should be identical

Comment (Walingar) [reply src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Why do you think so? I can add information about 17 May before 22 May.

**Review by Walingar — CHANGES_REQUESTED** [review]
A lot of code-style issues

File: src/main/java/impl/weather/YearTemperatureStatsImpl.java (hunk 37 above)
Comment (ammarali32) [reply src/main/java/impl/weather/YearTemperatureStatsImpl.java]: Ahh ok got it I thought it is not allowed i will fix it thanks

File: src/main/java/impl/weather/MonthInfo.java [hunk 39]
```diff
@@ -0,0 +1,52 @@
+package impl.weather;
+
+import api.weather.DayTemperatureInfo;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class MonthInfo {
+    private final Map < Integer, DayTemperatureInfo > info = new LinkedHashMap < > ();
```
Comment (Walingar) [inline src/main/java/impl/weather/MonthInfo.java]: code-style!!!! I won't close it but it is the last try. I told you: just use "Reformat code" action. It will help in 99% cases

**Review by Walingar — CHANGES_REQUESTED** [review]

**Review by Walingar — APPROVED** [review]
